### ISSUE TITLE
‘unable to get local issuer certificate’ when following express-openid-connect- tutorial

### ISSUE THREAD

**Issue by samlevin** [auth0/express-openid-connect#183]
### Describe the problem

When attempting to follow the tutorial outlined here https://auth0.com/blog/auth0-s-express-openid-connect-sdk my express app throws an unhandled 'unable to get local issuer certificate' exception even though everything appears to be configured correctly

### What was the expected behavior?

when accessing https://localhost/login I expected to see an auth screen (login form)

### Reproduction
- Follow the tutorial, create an auth0 app, ensure everything is wired up per the article
- Start caddy
- Start my local app
- Ensure settings match (OIDC info is correct)
- attempt to hit https://localhost/login
- app hangs, proxy eventually responds, app crashes due to an SSL error

### Environment

- **Version of this library used:** 2.2.1
- **Which framework are you using, if applicable:** Node.js v12.19.0
- **Other modules/plugins/libraries that might be involved:**
- **Any other relevant information you think would be useful:**

Full error trace:
```
    (node:48241) UnhandledPromiseRejectionWarning: AggregateError:
        RequestError: unable to get local issuer certificate
            at ClientRequest.<anonymous> (/Users/user/random-dev/api/node_modules/openid-client/node_modules/got/dist/source/core/index.js:953:111)
            at ClientRequest.origin.emit (/Users/user/random-dev/api/node_modules/openid-client/node_modules/@szmarczak/http-timer/dist/source/index.js:39:20)
        RequestError: unable to get local issuer certificate
            at ClientRequest.<anonymous> (/Users/user/random-dev/api/node_modules/openid-client/node_modules/got/dist/source/core/index.js:953:111)
            at ClientRequest.origin.emit (/Users/user/random-dev/api/node_modules/openid-client/node_modules/@szmarczak/http-timer/dist/source/index.js:39:20)
        at maybeSettle (/Users/user/random-dev/api/node_modules/p-some/index.js:31:11)
        at /Users/user/random-dev/api/node_modules/p-some/index.js:69:23
        at processTicksAndRejections (internal/process/task_queues.js:97:5)
    (node:48241) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 2)
```

**Comment by adamjmcgrath** [comment]
Hi @samlevin - 

That error looks like it's happening when the `openid-client` is making a request to your issuer's well-known endpoint. Node is telling you that it doesn’t have a trusted certificate authority installed for the certificate that your `issuerBaseURL` gave it.

**Couple of things to check:**
Is your machine running behind a proxy? Does Node trust the certificate that your proxy is issuing?
What is `issuerBaseURL`, is it giving out a trusted certificate? 
Can you do:
```js
const got = require('got');
(async () => {
  const response = await got(`${issuerBaseURL}/.well-known/openid-configuration`);
  console.log(response.body);
})();
```

**Comment by samlevin** [comment]
hey @adamjmcgrath, thanks for the reply!

1) My machine isn't running behind a proxy. My node.js instance, however, is running behind a proxy called Caddy (which is part of the official Auth0 environment setup guide, and is needed because TLS is required for the token endpoint authorization method):

I'm not sure if node.js trusts the certificate authority, but the guide above shows nothing beyond running a single command (`caddy reverse-proxy --from localhost:443 --to localhost:3000`). Apparently Caddy is supposed to be responsible for handling everything else transparently? (I could be mistaken here)

2) the output from that query can be found below:

```
{
	"issuer": "https://dev-zm9rci09.us.auth0.com/",
	"authorization_endpoint": "https://dev-zm9rci09.us.auth0.com/authorize",
	"token_endpoint": "https://dev-zm9rci09.us.auth0.com/oauth/token",
	"device_authorization_endpoint": "https://dev-zm9rci09.us.auth0.com/oauth/device/code",
	"userinfo_endpoint": "https://dev-zm9rci09.us.auth0.com/userinfo",
	"mfa_challenge_endpoint": "https://dev-zm9rci09.us.auth0.com/mfa/challenge",
	"jwks_uri": "https://dev-zm9rci09.us.auth0.com/.well-known/jwks.json",
	"registration_endpoint": "https://dev-zm9rci09.us.auth0.com/oidc/register",
	"revocation_endpoint": "https://dev-zm9rci09.us.auth0.com/oauth/revoke",
	"scopes_supported": ["openid", "profile", "offline_access", "name", "given_name", "family_name", "nickname", "email", "email_verified", "picture", "created_at", "identities", "phone", "address"],
	"response_types_supported": ["code", "token", "id_token", "code token", "code id_token", "token id_token", "code token id_token"],
	"code_challenge_methods_supported": ["S256", "plain"],
	"response_modes_supported": ["query", "fragment", "form_post"],
	"subject_types_supported": ["public"],
	"id_token_signing_alg_values_supported": ["HS256", "RS256"],
	"token_endpoint_auth_methods_supported": ["client_secret_basic", "client_secret_post"],
	"claims_supported": ["aud", "auth_time", "created_at", "email", "email_verified", "exp", "family_name", "given_name", "iat", "identities", "iss", "name", "nickname", "phone_number", "picture", "sub"],
	"request_uri_parameter_supported": false
}
```

**Comment by adamjmcgrath** [comment]
Hi @samlevin 

I don't think it's an issue with Caddy because it's not the request your browser is making to `https://localhost` that is failing, your error indicates that it is the request to `https://dev-zm9rci09.us.auth0.com/.well-known/openid-configuration` that is failing.

 > at ClientRequest.<anonymous> (/Users/user/random-dev/api/node_modules/openid-client/node_modules/got/dist/source/core/index.js:953:111)

This is the `openid-client` making a request from your node instance to `https://dev-zm9rci09.us.auth0.com/.well-known/openid-configuration` that is failing with `unable to get local issuer certificate`. 

This suggests to me that your node instance doesn't trust `https://dev-zm9rci09.us.auth0.com` - which is why I thought you might have a proxy on your machine (like ZScaler) that is causing the issue.

But when you call `await got(`${issuerBaseURL}/.well-known/openid-configuration`)` separately it succeeds, so I'm not sure.

**Comment by samlevin** [comment]
@adamjmcgrath I'm going to close this for now it appears to be (finally) consistently working, though admittedly I haven't changed much. thanks for weighing in.

**Comment by samlevin** [comment]
hey @adamjmcgrath,

For whatever reason, this keeps popping up off and on. So I took another look at your comment to try and grok what could be going wrong. I also created a new app (just incase).

The URL you had me hit before works as expected even for the new app. Yet, when I copy the auth0 sample configuration over to my node.js project and attempt to start a server + hit a login/logout endpoint, the same failure `unable to get local issuer certificate`.

I dove a bit deeper - started the app with `NODE_TLS_REJECT_UNAUTHORIZED=0` set, and then realized that the underlying OIDC client isn't actually attempting to hit the URL you had me test. Instead, it's hitting:
```
    RequestError: Unexpected token < in JSON at position 0 in "https://omitted.auth0.com/.well-known/oauth-authorization-server"
```

`https://omitted.auth0.com/.well-known/oauth-authorization-server` does not exist, the request does not work as expected, and `openid-client` throws an exception that `express-openid-connect` does not handle
`https://omitted.auth0.com/.well-known/openid-configuration` - however - works as expected in my browser, from cURL, via the snippet you shared - but the `openid-client` isn't hitting it (or maybe it is, but the exception from the other call side effects it?)

[Apparently there is a setting in `openid-client` which controls this?](https://github.com/panva/node-openid-client/blob/main/docs/README.md#issuerdiscoverissuer)

The maddening thing is that this has worked for me a few times through various changes/tweaks to settings (hail mary stuff). Is it possible that my configuration reverted back to something which may be causing the library to request the above instead of what you sent me? I can't imagine that auth0 has been sporadically 'down' for me over the course of 5+ days. It seems like either something in my environment or something in the library itself (possibly in response to something in my env)

relevant library config:
```
  const config = {
    authRequired: false,
    auth0Logout: true,
    secret: 'a long, randomly-generated string stored in env',
    baseURL: 'https://localhost',
    clientID: 'omitted',
    issuerBaseURL: 'https://omitted.auth0.com'
  };

  app.use(auth(config));
```

thanks again for your help

**Comment by adamjmcgrath** [comment]
> https://omitted.auth0.com/.well-known/oauth-authorization-server does not exist, the request does not work as expected, and openid-client throws an exception that express-openid-connect does not handle
https://omitted.auth0.com/.well-known/openid-configuration - however - works as expected in my browser, from cURL, via the snippet you shared - but the openid-client isn't hitting it (or maybe it is, but the exception from the other call side effects it?)

The `openid-client` attempts to hit both those urls, it doesn't matter if one of them fails as long as one of them succeeds. [pAny](https://github.com/panva/node-openid-client/blob/main/lib/issuer.js#L255) will succeed if any of the promises passed to it will succeed.

Your original error is a aggregate of 2 errors, meaning both requests have failed:

```
RequestError: unable to get local issuer certificate
...
RequestError: unable to get local issuer certificate
```

So basically, don't worry that `.well-known/oauth-authorization-server` fails, it's the `.well-known/openid-configuration` we're concerned with.

> I can't imagine that auth0 has been sporadically 'down' for me over the course of 5+ days. It seems like either something in my environment or something in the library itself (possibly in response to something in my env)

Yep, the only reason why the following code: https://github.com/panva/node-openid-client/blob/main/lib/issuer.js#L257-L261 would fail with `unable to get local issuer certificate` in the SDK and not when you run it directly using https://github.com/auth0/express-openid-connect/issues/183#issuecomment-768954708 that I can think of would be different proxy settings, you might have an https_proxy environment variable set?

**Comment by samlevin** [comment]
@adamjmcgrath thanks again for following up. unfortunately, this intermittent issue persists and I had to revert back to `passport-auth0` despite all of the evangelism for this new library. which is unfortunate because it was leaps and bounds easier to set up. I haven't had any type of issue since going the passport route.

To answer your question - this is a personal machine, with no proxy, VPN connection, or anything of the sort. I checked (double, triple) that `HTTPS_PROXY` is not set. 

I totally get where you're coming from, and agree that it could be something on my end. but to be crystal clear, sometimes this happens when I was running just the most basic example from the auth0 console. If I continue down the auth0 path then I'll probably have to try again at some point. any last ditch ideas as to what it could be?

**Comment by adamjmcgrath** [comment]
Hi @samlevin - could you try setting `baseURL: 'http://localhost:3000',` and just running the example on http.

You don't strictly need to use https on localhost (for Chrome at least), see this section from the tutorial: https://auth0.com/blog/auth0-s-express-openid-connect-sdk

> I could have easily skipped the proxy part, but given that the SameSite changes make HTTPS mandatory for SameSite=none cookies (see here for more context), it’s a good habit to start using HTTPS at development time anyway.

**Comment by madaster97** [comment]
@samlevin @adamjmcgrath , the `node-openid-client` library won't make the request to the OAuth server if you just specify the full URL in config.

Try setting this `baseURL: 'https://localhost/.well-known/openid-configuration'`. That should fix the issue (since the client will only request openid and not OAuth).

**Comment by adamjmcgrath** [comment]
Hi @madaster97 - thanks for your suggestion

The openid-client attempts to hit 2 urls, but it doesn't matter if one of them fails as long as one of them succeeds. So changing `issuerBaseURL` to the fully qualified url shouldn't make a difference - but @samlevin - please let me know if I'm wrong and this resolves your issue.

**Comment by adamjmcgrath** [comment]
Hi @samlevin - closing due to innactivity

Feel free to ping me if you have any other questions and I'll reopen

**Comment by synthmusic** [comment]
Hello, found this thread because I am facing the same issues.  I've followed along all the steps above and done some debugging as well, and I am thinking it is related to a thread in the Got project, in that it's definitely happening in the Got call for me, though it appears not to be the 'fault' of the got project because of other packages, also in thread: (https://github.com/sindresorhus/got/issues/951).

It also fails when running behind a reverse proxy (Caddy) or w/ a self issued cert running directly in express, though with different errors, as would e expected
`AggregateError: Issuer.discover() failed.
    RequestError: unable to get local issuer certificate
    RequestError: unable to get local issuer certificate`
`AggregateError: Issuer.discover() failed.
    RequestError: self signed certificate
    RequestError: self signed certificate`

also ran the Got request as suggested above and ran it with node-fetch for reference:
`        fetch(`${config.issuerBaseURL}/.well-known/openid-configuration`)
            .then((res) => res.text())
            .then((text) => console.log(text));

        const response = await got(`${config.issuerBaseURL}/.well-known/openid-configuration`);
        console.log(response.body);`

fetch successfully returns the config doc, and Got fails with the same cert error as above.

local details, 
"node": "16.15.1"
"express-openid-connect": "^2.7.3"
Win 10, however,  and lastly, I deployed to Heroku (which runs linux), and the error is now:  RequestError: connect ECONNREFUSED 127.0.0.1:443.  Got is definitely selecting localhost instead of the provided url.  the question is then why.
